### ISSUE TITLE
Error injecting fs in pytest class method + unittest.Testcase

### ISSUE THREAD

**Issue by staticdev** [jmcgeheeiv/pyfakefs#534]
**Describe the bug**
I tried injecting `fs` in a method of a class using unittest.Testcase + pytest but it did not work. 

**How To Reproduce**

Tested code:
```
# pdf_splitter.py
import os
import sys
import tempfile
import PyPDF4

class PdfSplitter:
    def __init__(self, filepath: str) -> None:
        self.filepath = filepath
        self.input_pdf = PyPDF4.PdfFileReader(filepath, "rb")
        self.total_pages = self.input_pdf.getNumPages()
        self.size = os.path.getsize(filepath)
        self.avg_size = self.size / self.total_pages

    def _get_pdf_size(self, pdf_writer: PyPDF4.PdfFileWriter) -> int:
        """Generates temporary PDF.

        Args:
            pdf_writer: pdf writer.

        Returns:
            int: generated file size.
        """
        with tempfile.TemporaryFile(mode="wb") as fp:
            pdf_writer.write(fp)
            return fp.tell()

    def split_max_size(self, max_size: int) -> int:
        """Creates new files based on max size.

        Args:
            max_size: size in integer megabytes.

        Returns:
            int: number of PDFs created.
        """
        if self.size > max_size:
            avg_step = int(max_size / self.avg_size)
            pdfs_count = 0
            current_page = 0

            while current_page != self.total_pages:
                end_page = current_page + avg_step
                if end_page > self.total_pages:
                    end_page = self.total_pages

                current_size = sys.maxsize
                # while PDF is too big create smaller PDFs
                while current_size > max_size:
                    pdf_writer = PyPDF4.PdfFileWriter()
                    for page in range(current_page, end_page):
                        pdf_writer.addPage(self.input_pdf.getPage(page))
                    current_size = self._get_pdf_size(pdf_writer)
                    self.input_pdf = PyPDF4.PdfFileReader(self.filepath, "rb")
                    end_page -= 1

                # write PDF with size max_size
                with open(
                    self.filepath.replace(".pdf", "-{}.pdf".format(pdfs_count)), "wb"
                ) as out:
                    pdf_writer.write(out)

                current_page = end_page + 1
                pdfs_count += 1
            return pdfs_count
        return 0
```

Testing code:
```
# test_pdf_splitter.py
import unittest
from unittest.mock import patch
import pytest
from pdf_split_tool import pdf_splitter

@patch("PyPDF4.PdfFileReader")
@patch("os.path.getsize")
class PdfSplitterTests(unittest.TestCase):
    @patch("PyPDF4.PdfFileWriter")
    @patch("tempfile.TemporaryFile")
    def test_split_max_size_greater_than_max(
        self,
        fs: Mock,
        mock_temporaryfile: Mock,
        mock_pypdf4_pdffilewriter: Mock,
        mock_os_path_getsize: Mock,
        mock_pypdf4_pdffilereader: Mock,
    ) -> None:
        # 2mb pdf with 1mb max_size
        mock_os_path_getsize.return_value = 2000000
        mock_pypdf4_pdffilereader.return_value.getNumPages.return_value = 2
        max_size = 1000000
        mock_temporaryfile.return_value.__enter__.return_value.tell.return_value = (
            1000000
        )
        splitter = pdf_splitter.PdfSplitter("big.pdf")
        assert splitter.split_max_size(max_size) == 2
```

I tried just injecting `fs` as in the [documentation](http://jmcgeheeiv.github.io/pyfakefs/release/usage.html#patch-using-the-pytest-plugin), also tried `@patchfs` and [`@pytest.fixture(autouse=True)`](https://docs.pytest.org/en/stable/unittest.html#using-autouse-fixtures-and-accessing-other-fixtures). I couldn't make it work.

Full code can be seen [here](https://github.com/staticdev/pdf-split-tool/blob/master/tests/test_pdf_splitter.py).

**Your enviroment**
```bash
python -c "import platform; print(platform.platform())"
>> Linux-5.4.0-39-generic-x86_64-with-glibc2.29
python -c "import sys; print('Python', sys.version)"
>> Python 3.8.2 (default, Apr  5 2020, 18:58:01) 
>> [GCC 9.2.1 20191008]
python -c "from pyfakefs.fake_filesystem import __version__; print('pyfakefs', __version__)"
>> pyfakefs 4.0.2
```

**Comment by mrbean-bremen** [comment]
In your testing example you use `unittest`, so you can either use `setUpPyfakefs` in your `setUp` function, or use `@patchfs`.
This works in the sense that the fake fs is used, but your test fails because the file in the fake filesystem is not found. If you use the fake fs, you have to work with it. Here is a variant of the example above that will work:
```py
@patch("PyPDF4.PdfFileReader")
class PdfSplitterTests(unittest.TestCase):
    @patch("PyPDF4.PdfFileWriter")
    # @patch("tempfile.TemporaryFile")
    @patchfs
    def test_split_max_size_greater_than_max(
            self,
            mock_pypdf4_pdffilewriter: Mock,
            mock_pypdf4_pdffilereader: Mock,
            fs
    ) -> None:
        # 2mb pdf with 1mb max_size
        mock_pypdf4_pdffilereader.return_value.getNumPages.return_value = 2
        max_size = 1000000
        fs.create_file("big.pdf", st_size=2000000)
        splitter = pdf_splitter.PdfSplitter("big.pdf")
        assert splitter.split_max_size(max_size) == 2
```
Note that I have removed the patches for `os.path.getsize` and `tempfile.TemporaryFile` - both work in the fake filesystem, and instead of mocking the file size, you create a fake file with that size.

**Comment by mrbean-bremen** [comment]
As a side note: pytest fixtures do not work with `unitest.TestCase` methods - if you want to use the pytest `fs` fixture, you could use instead:
```py
@patch("PyPDF4.PdfFileReader")
class TestPdfSplitter:
    @patch("PyPDF4.PdfFileWriter")
    def test_split_max_size_greater_than_max(
            self,
            mock_pypdf4_pdffilewriter: Mock,
            mock_pypdf4_pdffilereader: Mock,
            fs
    ) -> None:
    ...
```
Note that the test class is now not derived from `TestCase`, the name starts with `Test` (otherwise pytest won't find it), and instead of the `fs` parameter from `@patchfs` the `fs` fixture is used.

**Comment by staticdev** [comment]
@mrbean-bremen thanks for the quick response, but I am still having errors even with the changes your sent me:

```
__________________________________________________________________________ TestPdfSplitter.test_split_max_size_greater_than_max ___________________________________________________________________________

args = (<tests.test_pdf_splitter.TestPdfSplitter testMethod=test_split_max_size_greater_than_max>,), keywargs = {}
newargs = (<tests.test_pdf_splitter.TestPdfSplitter testMethod=test_split_max_size_greater_than_max>, <MagicMock name='PdfFileWriter' id='140661187677200'>, <MagicMock name='PdfFileReader' id='140661187670416'>)
newkeywargs = {}

    @wraps(func)
    def patched(*args, **keywargs):
        with self.decoration_helper(patched,
                                    args,
                                    keywargs) as (newargs, newkeywargs):
>           return func(*newargs, **newkeywargs)
E           TypeError: test_split_max_size_greater_than_max() missing 1 required positional argument: 'fs'

../../.pyenv/versions/3.8.2/lib/python3.8/unittest/mock.py:1348: TypeError
========================================================================================= short test summary info =========================================================================================
FAILED tests/test_pdf_splitter.py::TestPdfSplitter::test_split_max_size_greater_than_max - TypeError: test_split_max_size_greater_than_max() missing 1 required positional argument: 'fs'
```

**Comment by mrbean-bremen** [comment]
You did put `fs` as the last argument, right? Which version did you use?

**Comment by staticdev** [comment]
Yes I did. I am using pyfakefs 4.0.2.

**Comment by mrbean-bremen** [comment]
Thanks I:ll have another look tomorrow. By "which version" I actually meat which ox the versions I posted above, e.g. the unittest or the pytest version - I guess the first one.

**Comment by mrbean-bremen** [comment]
Hm, I tested both variants as shown above, and both work for me. While I tested under Windows, this shouldn't make a difference. Did you really use the exact same code?
I will check your original test code linked in your description later today, to check if there are more problems.

**Comment by mrbean-bremen** [comment]
I have adapted your tests accordingly and made PRs to show the different usages (using `setUpPyfakefs`, `@patchfs` and the `fs` fixture - for illustration only, delete them once you had a look) - and while the mypy tests fail (I didn't pay attention), the actual tests pass. 
So I'm still unclear what the problem is.

**Comment by staticdev** [comment]
@mrbean-bremen thanks a lot! I will review everything now =)

**Comment by staticdev** [comment]
After reviewing your PRs, I saw that the main problem was I was using `TestPdfSplitter(unittest.TestCase)`, so that I needed `@patchfs`. And there were also the options of using pyfakefs `TestCase` and using no TestCase class. I went for the last approach (https://github.com/staticdev/pdf-split-tool/pull/83) which also had smaller code and I found more general for using in other cases.

Thanks again for everything, closing this one.
